pkg/volumeclaimbinder/persistent_volume_claim_binder.go

```diff
@@ -0,0 +1,301 @@
+			glog.V(5).Infof("PersistentVolume[%s] is bound to PersistentVolumeClaim[%s]\n", volume.Name, volume.Spec.ClaimRef.Name)
+			// rebuild the Claim's Status as needed
+			if claim.Status.VolumeRef == nil {
+				syncClaimStatus(binderClient, volume, claim)
+			}
+		} else {
+			//claim was deleted by user.
+			glog.V(3).Infof("PersistentVolumeClaim[%s] unbound from PersistentVolume[%s]\n", volume.Spec.ClaimRef.Name, volume.Name)
+			// volume.Spec.ClaimRef is deliberately left non-nil so that another process can recycle the newly released volume
+			volume.Status.Phase = api.VolumeReleased
+			volume, err = binderClient.UpdatePersistentVolumeStatus(volume)
+			if err != nil {
+				return fmt.Errorf("Error updating pv: %+v\n", err)
+			}
+		}
+	} else {
+		volume.Status.Phase = api.VolumeAvailable
+		_, err := binderClient.UpdatePersistentVolumeStatus(volume)
+		if err != nil {
+			return fmt.Errorf("Error updating pv.status: %v\n", err)
+		}
+	}
+	return nil
+}
+
+func syncClaimStatus(binderClient binderClient, volume *api.PersistentVolume, claim *api.PersistentVolumeClaim) (err error) {
+	volumeRef, err := api.GetReference(volume)
+	if err != nil {
+		return fmt.Errorf("Unexpected error getting volume reference: %v\n", err)
+	}
+	// all "actuals" are transferred from PV to PVC so the user knows what
+	// type of volume they actually got for their claim
+	claim.Status.Phase = api.ClaimBound
+	claim.Status.VolumeRef = volumeRef
+	claim.Status.AccessModes = volume.Spec.AccessModes
+	claim.Status.Capacity = volume.Spec.Capacity
+
+	_, err = binderClient.UpdatePersistentVolumeClaimStatus(claim)
+
+	if err != nil {
+		claim.Status.Phase = api.ClaimPending
+		claim.Status.VolumeRef = nil
+		claim.Status.AccessModes = nil
+		claim.Status.Capacity = nil
+	}
+
+	return err
+}
+
+func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderClient, claim *api.PersistentVolumeClaim) (err error) {
+	glog.V(5).Infof("Synchronizing PersistentVolumeClaim[%s]\n", claim.Name)
+
+	if claim.Status.VolumeRef != nil {
+		glog.V(5).Infof("PersistentVolumeClaim[%s] is bound to PersistentVolume[%s]\n", claim.Name, claim.Status.VolumeRef.Name)
+		return nil
+	}
+
+	volume, err := volumeIndex.FindBestMatchForClaim(claim)
+	if err != nil {
+		return err
+	}
+
+	if volume != nil {
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
+		}
+
+		// make a binding reference to the claim
+		volume.Spec.ClaimRef = claimRef
+		volume, err = binderClient.UpdatePersistentVolume(volume)
+
+		if err != nil {
+			// volume no longer bound
+			volume.Spec.ClaimRef = nil
+			return fmt.Errorf("Error updating volume: %+v\n", err)
+		} else {
+			err = syncClaimStatus(binderClient, volume, claim)
+			if err != nil {
+				return fmt.Errorf("Error update claim.status: %+v\n", err)
```

Comment:
The rollback happens in the same syncClaimStatus func, because I use that func twice and didn't want to duplicate the rollback.